courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py

```diff
@@ -0,0 +1,79 @@
+"""Utilities for preprocessing natural language for machine learning"""
+
+import unicodedata
+import re
+
+import tensorflow as tf
+
+def unicode_to_ascii(s):
+        return ''.join(c for c in unicodedata.normalize('NFD', s)
+            if unicodedata.category(c) != 'Mn')
+
+def preprocess_sentence(w):
+    w = unicode_to_ascii(w.lower().strip())
+
+    # creating a space between a word and the punctuation following it
+    # eg: "he is a boy." => "he is a boy ."
+
+    w = re.sub(r"([?.!,¿])", r" \1 ", w)
```

Comment:
* It seems that the documentation is not in sync with the regex. Shouldn't it be rather:  'Surrounding punctuations with a single space: "he is a boy." => "he is a boy . " ' If it's not he behavior you want, then the regex maybe more something like: r" \1" (i.e. no space after group \1)
* Why do we want to treat other punctuation signs differently like ";:,"? Have they been stripped already?